### DISCUSSION TITLE
Weekly updated open sourced model implementations in Flax

### DISCUSSION THREAD

**Discussion by DarshanDeshpande** [google/flax#1770] [Show and tell]
JAX Models repository aims to provide open sourced JAX/Flax implementations for research papers originally without code or code written with frameworks other than JAX. The goal of this project is to make a collection of models, layers, activations and other utilities that are most commonly used for research

Some features of jax_models:
1. Pip installable
2. Weekly / fortnightly model additions
3. Comprehensive tests
4. Model architectures are cross checked by checking original implementations if available or otherwise miscellaneous aspects of the paper such as model parameters

The repository is currently in its pre-alpha stage with only around 6 model and layer implementations. If you are interested in this, you can contribute by either asking for specific paper implementations or by helping me train and open source weights of these models.

If you like my work then please consider starring the repository and giving feedback. Thank you!

Link: https://github.com/DarshanDeshpande/jax-models

**Comment by avital** [comment]
Thanks for sharing @DarshanDeshpande! Now that you've implemented these models, please let us know if there were challenges you hit, so that we can improve Flax going forward for others!

**Reply by DarshanDeshpande** [reply]
There's nothing as of now. Flax is really mature and flexible as it is. 

If you could just direct me to some good benchmarks comparing Flax with other frameworks like TF or Pytorch, it would be very helpful. All the ones that I keep finding aren't comprehensive enough :(

**Comment by andsteing** [comment]
See also @matthias-wright's repository of pretrained models (mentioned in #1320):
https://github.com/matthias-wright/flaxmodels

When re-implementing models it might be particularly useful to provide code for loading PyTorch weights and adding tests verifying that both implementations produce identical results.

**Reply by DarshanDeshpande** [reply]
I thought about porting weights but it's a bit difficult to do than it seemed originally. I have successfully ported weights from Tensorflow to JAX and vice versa but for torch, it becomes a bit more complicated. For example, the default convolution data format for torch is NCHW whereas for both TF and JAX it's NHWC. So for cases like these, most torch codes end up sandwiching convolutions between transpose operations which makes it equivalent to the default TF/JAX NHWC implementation while at the same time, making it a bit inconvenient to port. Though still possible, I am still trying to weigh the pros and cons here. Let me know if you have any opinion about this or if you know how to handle such differences effectively.

**Reply by andsteing** [reply]
Yeah `NCHW` vs. `NCHW` convolution order is definitely a gotcha. There are some more about slightly different implementations of activation functions IIRC.

Maybe @matthias-wright can chime in since he already went through the work of porting some of the models in the https://github.com/matthias-wright/flaxmodels repo.

Ideally we would have a guide that explains the various pitfalls when porting models PT->JAX and TF->JAX. Pull requests welcome :)

**Reply by matthias-wright** [reply]
Hi everyone, here are some things off the top of my head:

* As mentioned before, PyTorch uses NCHW while Jax/Flax/TF use NHWC. This is not a problem for porting weights though. You just have to transpose the weights accordingly.
* However, you have to be careful, when you have a model that uses convolutions followed by some fc layers (ResNet, VGG, etc). After the convolutions, the activations of shape [N, C, H, W] are usually reshaped to [N, C * H * W] and then fed to the fc layers. When you port your weights from PyToch to Flax, your activations after the convolutions will be [N, H, W, C]. Before you reshape the activations for the fc layers, you have to transpose them to [N, C, H, W] (see [here](https://github.com/matthias-wright/flaxmodels/blob/b8025949472e4fb4fc34800b9f5b03ae639c9ff2/flaxmodels/vgg/vgg.py#L127)).
* The kernel of `torch.nn.Linear` has dimensions [out_features, in_features], while the kernel of `flax.linen.Dense` has dimensions [in_features, out_features]. As before, transposing the kernel will do the trick.
* `torch.nn.BatchNorm2d` uses `0.1` as the default value for the `momentum` parameter while `flax.linen.BatchNorm` uses `0.9`. This corresponds to the same computation, because PyTorch multiplies the estimated statistic with `(1−momentum)` and the new observed value with `momentum`, while Flax multiplies the estimated statistic with `momentum` and the new observed value with `(1−momentum)`. I once made the mistake of setting `momentum=0.1` in Flax because the PyTorch model used `momentum=0.1`. Things like that are super difficult to catch.
* `torch.nn.ConvTranspose2d` and `flax.linen.ConvTranspose` are not the same. `flax.linen.ConvTranspose` is a wrapper around `jax.lax.conv_transpose` which computes a fractionally strided convolution, while `torch.nn.ConvTranspose2d` computes a gradient based transposed convolution.
* `torch.nn.AvgPool2d` has an option `count_include_pad`. If `count_include_pad=True`, the padded zeros will be included in the average computation. The default value is `True`. As long as `count_include_pad=True` in PyTorch, `torch.nn.AvgPool2d` and `flax.linen.avg_pool` are compatible. However, when `count_include_pad=False` in `torch.nn.AvgPool2d`, it will produce different results than `flax.linen.avg_pool`. I implemented a wrapper around `flax.linen.avg_pool` that works like `torch.nn.AvgPool2d` with `count_include_pad=False` [here](https://github.com/matthias-wright/jax-fid/blob/main/jax_fid/inception.py#L625).

Hope that helps!

**Reply by DarshanDeshpande** [reply]
This was extremely helpful. Thank you so much! @matthias-wright

**Reply by marcvanzee** [reply]
Wow, this is really near @matthias-wright, thanks a lot for sharing! This will really help the community a lot, and it is amazing you figured this all out!

I think we should  publicize this more broadly, maybe in some "How to convert Torch models to Flax" guide, because I am sure this is useful to the community in general, and it is a bit hidden away now in this discussion.

Would you be interested in writing this down as a HOWTO? I don't think you have to make it much more elaborate as you did now, but it might be useful if for each item you could point to some code snippet.

**Reply by matthias-wright** [reply]
Sounds good @marcvanzee! Are you thinking about something that looks like [this](https://flax.readthedocs.io/en/latest/howtos/model_surgery.html) ?

**Reply by marcvanzee** [reply]
Awesome! Yes, indeed, I was referring to a HOWTO, which we also sometimes call a pattern, tutorial, ...

We are still thinking how to best structure our guides, if you have ideas for how to improve this feel free to chime in as well.
In any case, for now a HOWTO seems like the right place.

**Reply by matthias-wright** [reply]
Okay cool, I will then submit a PR once I have something.

**Reply by andsteing** [reply]
This is now submitted as part of #1848 
https://flax.readthedocs.io/en/latest/howtos/convert_pytorch_to_flax.html

Thanks for the nice PR, Matthias!

**Reply by DarshanDeshpande** [reply]
@andsteing @matthias-wright Do you think it's a good idea for the tutorial to also mention the approximations made by some layers? For example: `LayerNorm` or the default `approximate=True` as seen for gelu?
